feat(cli): Added an `--all` flag to `meltano state clear` to delete all state IDs

fauzxan opened the pull request:
<!--

Please, go through these steps when you submit a PR.

1. Make sure your branch is not protected. In particular, avoid making PRs from the `main` branch of your fork.

2. Give a descriptive title to your PR. We use semantic titles, and the accepted types and scopes are listed in https://github.com/meltano/meltano/blob/main/.github/semantic.yml.

   A good title should look like this:

   ```
   feat(cli): The `meltano run` command now accepts a `--timeout` option to limit the time it runs
   ```

3. Provide a description of your changes.

4. Put "Closes #XXXX" in your comment to auto-close the issue that your PR fixes (if such).

-->

## Description

<!-- Describe the changes introduced by this PR -->
This PR introduces a feature allowing users to clear all states with the `--all` argument in `meltano state clear` command.

### Design

- `state.clear_state` to take in an extra argument `--all`.
- `state-id` argument has become optional. However, user is prompted to serve one of either `--all` or `state-id`, and not both.
- `state_service.clear_all_states` has been added that gets all `state_id`s and calls `state_store_manager.clear()` on all state_ids.
- `TestCliState.test_clear_all` added to test this new functionality

### To-Do:

- [x] add a default clear_all() implementation in the base StateStoreManage
Eventually, clear_all() will be implemented in all StateStoreManagers to bulk delete the states.
- [ ] Address errors in https://results.pre-commit.ci/run/github/378997214/1726756744.X8yggwgrQWKXwLLkifo0_w

## Related Issues

* Closes #8749

fauzxan committed:
add: new state service clear_all_states

fauzxan committed:
add: all argument + state service support

fauzxan committed:
update: make --all optional

fauzxan committed:
Merge branch 'main' into add-all-option-state-clear

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.ci hooks

for more information, see https://pre-commit.ci

edgarrmondragon commented on src/meltano/cli/state.py:
```diff
@@ -310,12 +310,26 @@ def get_state(ctx: click.Context, project: Project, state_id: str) -> None:
+        state_service.clear_state(state_id)
+    if all:
+        state_service: StateService = ctx.obj[STATE_SERVICE_KEY]
+        state_service.clear_all_states()
```
suggestion: should we emit a message here saying how many state IDs were cleared?

edgarrmondragon replied:
c71c195c8

edgarrmondragon commented on tests/meltano/cli/test_state.py:
```diff
@@ -281,6 +281,14 @@ def test_clear(self, state_service, cli_runner, state_ids) -> None:
+        with mock.patch("meltano.cli.state.StateService", return_value=state_service):
+            result = cli_runner.invoke(cli, ["state", "clear", "--force", "--all"])
+            assert_cli_runner(result)
+            for state_id in state_ids:
+                job_state = state_service.get_state(state_id)
+                assert (not job_state) or (not job_state.get("singer_state"))
```
suggestion: could we use `StateService.list_state` to check the before (> 0) and after (0) state ID count?

edgarrmondragon commented on src/meltano/cli/state.py:
```diff
@@ -310,12 +310,26 @@ def get_state(ctx: click.Context, project: Project, state_id: str) -> None:
-@click.argument("state-id")
+@click.argument("state-id", required=False)
+@click.option(
+    "--all",
```
change:

I'm a bit surprised Ruff is not complaining about this shadowing the builtin function `all`, but we should still change the parameter name:

```suggestion
    ["--all", "clear_all"],
```

fauzxan replied:
Ruff did complain [here](https://results.pre-commit.ci/run/github/378997214/1726756744.X8yggwgrQWKXwLLkifo0_w) haha!
Anyways, I changed this.

edgarrmondragon commented on src/meltano/cli/state.py:
```diff
@@ -310,12 +310,26 @@ def get_state(ctx: click.Context, project: Project, state_id: str) -> None:
+    "--all",
+    is_flag=True,
+    required=False,
+    help="Helps clear all states. Use with --force flag to skip confirmation.",
```
question: let me know if you plan to implement this. I think it's certainly worth it since it's a destructive action, but should it be something like `--yes/-y` instead?

edgarrmondragon replied:
Nevermind, this flag was already present.

edgarrmondragon commented on src/meltano/cli/state.py:
```diff
@@ -310,12 +310,26 @@ def get_state(ctx: click.Context, project: Project, state_id: str) -> None:
 @pass_project(migrate=True)
 @click.pass_context
-def clear_state(ctx: click.Context, project: Project, state_id: str) -> None:
+def clear_state(ctx: click.Context, project: Project, state_id: str, all: bool) -> None:
```
See above

```suggestion
def clear_state(ctx: click.Context, project: Project, state_id: str, clear_all: bool) -> None:
```

edgarrmondragon commented on src/meltano/cli/state.py:
```diff
@@ -310,12 +310,26 @@ def get_state(ctx: click.Context, project: Project, state_id: str) -> None:
-        state_service_from_state_id(project, state_id) or ctx.obj[STATE_SERVICE_KEY]
-    )
-    state_service.clear_state(state_id)
+    if (not state_id and not all) or (
+        state_id and all
+    ):  # Case where neither or both have been provided
```
```suggestion
    # Case where neither or both have been provided
    if bool(state_id) == clear_all:
```

fauzxan replied:
Updated!

edgarrmondragon commented on src/meltano/cli/state.py:
```diff
@@ -310,12 +310,26 @@ def get_state(ctx: click.Context, project: Project, state_id: str) -> None:
+    if (not state_id and not all) or (
+        state_id and all
+    ):  # Case where neither or both have been provided
+        raise click.UsageError("Either a state ID or --all flag must be provided.")
```
suggestion: either add test coverage for this. Or, since it's a fairly straightforward error state, don't count it

```suggestion
        raise click.UsageError("Either a state ID or --all flag must be provided.")  # pragma: no cover
```

fauzxan replied:
Done.

fauzxan committed:
fix: ruff suggestions

https://results.pre-commit.ci/run/github/378997214/1726756744.X8yggwgrQWKXwLLkifo0_w

fauzxan committed:
add: clear_all() abstract method

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.ci hooks

for more information, see https://pre-commit.ci

edgarrmondragon commented:
Thanks for getting this started @fauzxan!

fauzxan committed:
Merge branch 'main' into add-all-option-state-clear

fauzxan committed:
update: nits

- remove `Args:`
- change check condition
- remove UsageError message from qa cover

fauzxan committed:
Merge branch 'add-all-option-state-clear' of https://github.com/fauzxan/meltano into add-all-option-state-clear

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.ci hooks

for more information, see https://pre-commit.ci

fauzxan committed:
Merge branch 'main' into add-all-option-state-clear

edgarrmondragon commented on src/meltano/core/state_store/base.py:
```diff
@@ -69,6 +69,11 @@ def clear(self, state_id):  # noqa: ANN001, ANN201
+    @abstractmethod
+    def clear_all(self):  # noqa: ANN201
+        """Clear all states."""
+        ...
```
```suggestion
        for state_id in self.get_state_ids("*"):
            self.clear(state_id)
```

edgarrmondragon commented on src/meltano/core/state_store/base.py:
```diff
@@ -69,6 +69,11 @@ def clear(self, state_id):  # noqa: ANN001, ANN201
         """
         ...
 
+    @abstractmethod
```
```suggestion
```

edgarrmondragon commented on src/meltano/core/state_store/base.py:
```diff
@@ -69,6 +69,11 @@ def clear(self, state_id):  # noqa: ANN001, ANN201
         ...
 
+    @abstractmethod
+    def clear_all(self):  # noqa: ANN201
```
```suggestion
    def clear_all(self) -> None:
```

edgarrmondragon commented on src/meltano/core/state_service.py:
```diff
@@ -193,6 +193,11 @@ def clear_state(self, state_id, *, save: bool = True) -> None:  # noqa: ANN001,
 
+    def clear_all_states(self) -> None:
+        """Clear all states."""
+        for state_id in self.state_store_manager.get_state_ids("*"):
+            self.state_store_manager.clear(state_id)
```
```suggestion
        self.state_store_manager.clear_all(state_id)
```

edgarrmondragon commented:
@fauzxan let me know if you're still interested in moving forward with this PR

edgarrmondragon committed:
refactor: Move default implementation to

edgarrmondragon committed:
Merge branch 'main' into add-all-option-state-clear

edgarrmondragon commented on src/meltano/core/state_service.py:
```diff
@@ -193,6 +193,10 @@ def clear_state(self, state_id, *, save: bool = True) -> None:  # noqa: ANN001,
 
+    def clear_all_states(self) -> None:
+        """Clear all states."""
+        self.state_store_manager.clear_all(state_id)
```
```suggestion
        self.state_store_manager.clear_all()
```

edgarrmondragon committed:
Update src/meltano/core/state_service.py

edgarrmondragon commented on src/meltano/cli/state.py:
```diff
@@ -310,12 +310,28 @@ def get_state(ctx: click.Context, project: Project, state_id: str) -> None:
 @click.pass_context
-def clear_state(ctx: click.Context, project: Project, state_id: str) -> None:
+def clear_state(
+    ctx: click.Context, project: Project, state_id: str, clear_all: bool
```
```suggestion
    ctx: click.Context,
    project: Project,
    state_id: str,
    clear_all: bool,  # noqa: FBT001
```

edgarrmondragon commented on src/meltano/cli/state.py:
```diff
@@ -310,12 +310,28 @@ def get_state(ctx: click.Context, project: Project, state_id: str) -> None:
-        state_service_from_state_id(project, state_id) or ctx.obj[STATE_SERVICE_KEY]
-    )
-    state_service.clear_state(state_id)
+    if bool(state_id) == clear_all:  # Case where neither or both have been provided
+        raise click.UsageError(
+            "Either a state ID or --clear-all flag must be provided. Not both."
+        )  # pragma: no cover
```
```suggestion
    # Case where neither or both have been provided
    if bool(state_id) == clear_all:  # pragma: no cover
        msg = "Either a state ID or --clear-all flag must be provided, not both"
        raise click.UsageError(msg)
```

edgarrmondragon committed:
Apply suggestions from code review

edgarrmondragon committed:
Merge branch 'main' into add-all-option-state-clear

edgarrmondragon committed:
Test using `StateService.list_state`

edgarrmondragon committed:
Custom implementation for SQL state store

edgarrmondragon committed:
`--clear-all` -> `--all`

edgarrmondragon committed:
Print the count of deleted messages

edgarrmondragon committed:
Merge branch 'main' into add-all-option-state-clear

edgarrmondragon commented on src/meltano/cli/state.py:
```diff
@@ -309,13 +309,35 @@ def get_state(ctx: click.Context, project: Project, state_id: str) -> None:
+    "clear_all",
+    is_flag=True,
+    required=False,
+    help="Helps clear all states. Use with --force flag to skip confirmation.",
```
```suggestion
    help="Clear all states IDs.",
```

edgarrmondragon committed:
Update help message

edgarrmondragon commented on src/meltano/cli/state.py:
```diff
@@ -309,13 +309,35 @@ def get_state(ctx: click.Context, project: Project, state_id: str) -> None:
-    state_service.clear_state(state_id)
+    # Case where neither or both have been provided
+    if bool(state_id) == clear_all:  # pragma: no cover
+        msg = "Either a state ID or --clear-all flag must be provided, not both"
```
```suggestion
        msg = "Either a state ID or --all flag must be provided, not both"
```

edgarrmondragon committed:
Update error message

edgarrmondragon commented on src/meltano/cli/state.py:
```diff
@@ -309,13 +309,35 @@ def get_state(ctx: click.Context, project: Project, state_id: str) -> None:
-    state_service.clear_state(state_id)
+    # Case where neither or both have been provided
+    if bool(state_id) == clear_all:  # pragma: no cover
+        msg = "Either a state ID or --all flag must be provided, not both"
```
```suggestion
        msg = "Either a state ID or the --all flag must be provided, not both"
```

edgarrmondragon committed:
Update src/meltano/cli/state.py

edgarrmondragon committed:
Merge branch 'main' into add-all-option-state-clear

WillDaSilva commented on src/meltano/cli/state.py:
```diff
@@ -309,13 +309,35 @@ def get_state(ctx: click.Context, project: Project, state_id: str) -> None:
-    )
-    state_service.clear_state(state_id)
+    # Case where neither or both have been provided
+    if bool(state_id) == clear_all:  # pragma: no cover
```
This seems a bit clearer

```suggestion
    if state_id and clear_all:  # pragma: no cover
```

edgarrmondragon replied:
I've pushed 13e2e9eba1d434c75d5d0721dcd4044543bfee6e to actually test this code branch. The idea is that the sub-command should fail when the user

- provides both the `state-id` positional arg and the `--all` flag
- doesn't provide the `state-id` positional arg nor the `--all` flag

Does that make sense? This may be more readable?

```python
if (state_id is None and not clear_all) or (state_id is not None and clear_all):
```

WillDaSilva replied:
Ah right, this is an XNOR operation, and `a XNOR b` is equivalent to `bool(a) == bool(b)`. I'd say it's fine as-is.

WillDaSilva commented on src/meltano/cli/state.py:
```diff
@@ -309,13 +309,35 @@ def get_state(ctx: click.Context, project: Project, state_id: str) -> None:
+def clear_state(
+    ctx: click.Context,
+    project: Project,
+    state_id: str,
```
```suggestion
    state_id: str | None,
```

edgarrmondragon committed:
Update src/meltano/cli/state.py

Co-authored-by: Will Da Silva <will@willdasilva.xyz>

edgarrmondragon committed:
Test usage error when passed both a state ID and --all, or neither

edgarrmondragon committed:
Use the right type

edgarrmondragon commented:
@all-contributors
please add @fauzxan for code.
